compiler/rustc_mir_transform/src/lint_tail_expr_drop_order.rs

```diff
@@ -0,0 +1,382 @@
+        for (statement_index, stmt) in data.statements.iter().enumerate() {
+            if let StatementKind::BackwardIncompatibleDropHint { place, reason: _ } = &stmt.kind {
+                bid_per_block
+                    .entry(block)
+                    .or_insert(vec![])
+                    .push((Location { block, statement_index }, &**place));
+                bid_places.insert(&**place);
+            }
+        }
+    }
+    if bid_per_block.is_empty() {
+        return;
+    }
+
+    dump_mir(tcx, false, "lint_tail_expr_drop_order", &0 as _, body, |_, _| Ok(()));
+    let param_env = tcx.param_env(def_id).with_reveal_all_normalized(tcx);
+    let is_closure_like = tcx.is_closure_like(def_id.to_def_id());
+    let move_data = MoveData::gather_moves(body, tcx, param_env, |_| true);
+    let maybe_init = MaybeInitializedPlaces::new(tcx, body, &move_data);
+    let mut maybe_init =
+        maybe_init.into_engine(tcx, body).iterate_to_fixpoint().into_results_cursor(body);
+    for (&block, candidates) in &bid_per_block {
+        let mut all_locals_dropped = ChunkedBitSet::new_empty(move_data.move_paths.len());
+        let mut linted_candidates = BTreeSet::new();
+        let mut drop_span = None;
+        for (idx, &(candidate, place)) in candidates.iter().enumerate() {
+            maybe_init.seek_after_primary_effect(candidate);
+            let MaybeReachable::Reachable(maybe_init_in_future) = maybe_init.get() else {
+                continue;
+            };
+            debug!(maybe_init_in_future = ?maybe_init_in_future.iter().map(|idx| &move_data.move_paths[idx]).collect::<Vec<_>>());
+            let maybe_init_in_future = maybe_init_in_future.clone();
+            for block in drops_reachable_from_location(body, block, place).iter() {
+                let data = &body.basic_blocks[block];
+                if drop_span.is_none() {
+                    drop_span = data
+                        .terminator
+                        .as_ref()
+                        .map(|term| body.source_scopes[term.source_info.scope].span);
+                }
+                debug!(?candidate, "inspect");
+                maybe_init.seek_before_primary_effect(Location {
+                    block,
+                    statement_index: data.statements.len(),
+                });
+                let MaybeReachable::Reachable(maybe_init_now) = maybe_init.get() else { continue };
+                let mut locals_dropped = maybe_init_in_future.clone();
+                debug!(maybe_init_now = ?maybe_init_now.iter().map(|idx| &move_data.move_paths[idx]).collect::<Vec<_>>());
+                locals_dropped.subtract(maybe_init_now);
+                debug!(locals_dropped = ?locals_dropped.iter().map(|idx| &move_data.move_paths[idx]).collect::<Vec<_>>());
+                if all_locals_dropped.union(&locals_dropped) {
+                    linted_candidates.insert(idx);
+                }
+            }
+        }
+        {
```

Comment:
I'm a bit confused by this code... add a comment in front of this block with some code examples?

Reply:
Gladly. Examples are added in the comment.